File: apps/identity-wallet/src/lib/ipc/app-passwords.ts
import { invoke } from '@tauri-apps/api/core';
import { authenticateBiometric } from '$lib/biometric';
import type { UnlockReason } from './identity';

// ── App passwords ("Sign in to Bluesky and other apps") ───────────────────────
//
// A key-sovereign Custos account is passwordless, so the official Bluesky app —
// which signs into a third-party PDS with password `createSession`, not OAuth —
// has nothing to type. The app password minted here is that missing credential:
// a scoped, revocable password that opens a `com.atproto.appPass` session. That
// session can post, like, follow, and browse via the AppView, but can never touch
// account management, identity/PLC operations, agents, or app passwords themselves.
// Direct messages (chat) additionally require the `privileged` flag at mint time.

/** Result of minting an app password. `password` is shown ONCE — never retrievable again. */
export type AppPasswordCreated = {
  name: string;
  /** The generated `xxxx-xxxx-xxxx-xxxx` secret. Display once, offer copy, then drop it. */
  password: string;
  createdAt: string;
  privileged: boolean;
};

/** One existing app password — metadata only, never the secret. */
export type AppPasswordEntry = {
  name: string;
  createdAt: string;
  privileged: boolean;
};

/**
 * Error returned by the app-password commands.
 * Matches `AppPasswordsError` in `app_passwords.rs`
 * (`#[serde(tag = "code", rename_all = "SCREAMING_SNAKE_CASE", rename_all_fields = "camelCase")]`) —
 * codes must match exactly.
 *
 * `SESSION_LOCKED` is the cue to run the passwordless {@link sovereignLogin} (biometric) and retry.
 */
export type AppPasswordsError =
  // The identity is locked — run sovereignLogin(did) and retry. `reason` mirrors ensureIdentitySession.
  | { code: 'SESSION_LOCKED'; reason: UnlockReason }
  // The hosting PDS rate-limited the request; `retryAfter` is the server's Retry-After.
  | { code: 'RATE_LIMITED'; retryAfter: string | null }
  // An app password with this name already exists — pick a different name.
  | { code: 'DUPLICATE_NAME' }
  | { code: 'IDENTITY_NOT_FOUND'; message: string }
  // A server-side step failed for a non-connectivity reason. `status` is the HTTP code when
  // the hosting PDS refused the request, or null for a non-HTTP session failure (unsupported
  // host, keychain, malformed response); `message` is the server's own error text.
  | { code: 'SERVER_ERROR'; status: number | null; message: string }
  | { code: 'NETWORK_ERROR'; message: string };

/**
 * Mint a named app password for the identity. Resolves with the generated secret,
 * which is surfaced ONCE — the server stores only a hash.
 *
 * The biometric prompt precedes the IPC invocation: minting creates a durable login
 * credential for the account, so cancellation must reach neither Rust nor the network.
 * Set `privileged` only when the credential needs direct-message (chat) access.
 */
export const createAppPassword = async (
  did: string,
  name: string,
  privileged: boolean
): Promise<AppPasswordCreated> => {
  await authenticateBiometric('Create an app password for this identity');
  return invoke('create_app_password', { did, name, privileged });
};

/** List the identity's app passwords — names, creation times, privilege; never secrets. */
export const listAppPasswords = (did: string): Promise<AppPasswordEntry[]> =>
  invoke('list_app_passwords', { did });

/**
 * Revoke a named app password. The server deletes the credential and its
 * sessions/refresh tokens atomically, so a signed-in app is cut off immediately.
 * Callers gate this behind `authenticateBiometric()` in the confirming screen.
 */
export const revokeAppPassword = (did: string, name: string): Promise<void> =>
  invoke('revoke_app_password', { did, name });

File: crates/custos-client/src/app_passwords.rs
// pattern: Imperative Shell

//! App-password management: typed XRPC methods over an authenticated (full-access)
//! [`OAuthClient`] for `com.atproto.server.{create,list,revoke}AppPassword`.

use serde::{Deserialize, Serialize};

use crate::error::{xrpc_json, xrpc_ok, PdsClientError, TransportObserver};
use crate::oauth_client::OAuthClient;

/// Result of minting an app password (`com.atproto.server.createAppPassword`).
/// `password` is the generated secret, surfaced ONCE at creation — the server
/// stores only its hash, so it can never be retrieved again.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppPasswordCreated {
    pub name: String,
    /// The generated `xxxx-xxxx-xxxx-xxxx` secret. Shown once; never retrievable.
    pub password: String,
    pub created_at: String,
    pub privileged: bool,
    /// The Custos personal-details grant (ADR-0033). Defaults to `false` when the field is
    /// absent — which is exactly what a non-Custos PDS returns, so a requested-but-ignored
    /// grant reads back honestly as not granted.
    #[serde(default)]
    pub personal_details: bool,
}

/// One app-password entry from `listAppPasswords` — public metadata only, never the secret.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppPasswordEntry {
    pub name: String,
    pub created_at: String,
    pub privileged: bool,
    /// The Custos personal-details grant (ADR-0033); `false` when the host omits the field.
    #[serde(default)]
    pub personal_details: bool,
}

#[derive(Deserialize)]
struct ListAppPasswordsResponse {
    passwords: Vec<AppPasswordEntry>,
}

/// Mint a named app password on the hosting PDS.
///
/// Calls `POST /xrpc/com.atproto.server.createAppPassword`. Requires a full-access
/// session (an app-password session cannot mint more app passwords). A duplicate
/// name surfaces as `XrpcError { status: 409, .. }`.
pub async fn create_app_password(
    client: &OAuthClient,
    name: &str,
    privileged: bool,
    personal_details: bool,
    observer: &dyn TransportObserver,
) -> Result<AppPasswordCreated, PdsClientError> {
    let resp = client
        .post(
            "/xrpc/com.atproto.server.createAppPassword",
            &serde_json::json!({
                "name": name,
                "privileged": privileged,
                "personalDetails": personal_details,
            }),
        )
        .await
        .map_err(|e| PdsClientError::NetworkError {
            message: format!("create_app_password failed: {}", e),
        })?;

    xrpc_json("createAppPassword", resp, observer).await
}

/// List the account's app passwords (names, creation times, privilege — never secrets).
///
/// Calls `GET /xrpc/com.atproto.server.listAppPasswords`. Requires a full-access session.
pub async fn list_app_passwords(
    client: &OAuthClient,
    observer: &dyn TransportObserver,
) -> Result<Vec<AppPasswordEntry>, PdsClientError> {
    let resp = client
        .get("/xrpc/com.atproto.server.listAppPasswords")
        .await
        .map_err(|e| PdsClientError::NetworkError {
            message: format!("list_app_passwords failed: {}", e),
        })?;

    xrpc_json::<ListAppPasswordsResponse>("listAppPasswords", resp, observer)
        .await
        .map(|body| body.passwords)
}

/// Revoke a named app password (and, server-side, its sessions/refresh tokens atomically).
///
/// Calls `POST /xrpc/com.atproto.server.revokeAppPassword`. Idempotent on the server.
pub async fn revoke_app_password(
    client: &OAuthClient,
    name: &str,
    observer: &dyn TransportObserver,
) -> Result<(), PdsClientError> {
    let resp = client
        .post(
            "/xrpc/com.atproto.server.revokeAppPassword",
            &serde_json::json!({ "name": name }),
        )
        .await
        .map_err(|e| PdsClientError::NetworkError {
            message: format!("revoke_app_password failed: {}", e),
        })?;

    xrpc_ok("revokeAppPassword", resp, observer)
        .await
        .map(|_| ())
}
